deploy-agent/deployd/common/stats.py

```diff
@@ -28,69 +49,287 @@ def __exit__(self, stats, sample_rate=1.0, tags=None):
+        obj = json.loads(self.ins)
+        self.mtype = obj.get('mtype', None)
+        self.name = obj.get('name', None)
+        self.value = obj.get('value', None)
+        self.sample_rate = obj.get('sample_rate', None)
+        self.tags = obj.get('tags', None)
+
+    def deserialize(self, ins=None):
+        """ attempt to deserialize
+            :param: ins json as str
+            return: bool, False on error
+        """
+        if ins:
+            self.ins = ins
+        try:
+            self._deserialize()
+            return True
+        except self.JSONDecodeError:
+            return False
+        except TypeError:
+            return False
+
+
+class MetricClient:
+    """ metrics client wrapper, enables disk cache """
+
+    def __init__(self, port=METRIC_PORT_HEALTH, cache_path=METRIC_CACHE_PATH):
+        self.port = port
+        self.cache = MetricCache(path=cache_path)
+        self.stat = None
+
+    @staticmethod
+    def _add_default_tags(tags=None):
+        """ add default tags to stats
+            :param: tags as dict
+            return: dict
+        """
+        if not __version__:
+            # defensive case, should not be hit
+            return tags
+        if __version__ and not tags:
+            tags = dict()
+        tags['deploy_agent_version'] = __version__
+        return tags
+
+    @staticmethod
+    def _parse_stat(mtype=None, name=None, value=None, sample_rate=None, tags=None):
+        """ return Stat for given kwargs """
+        return Stat(mtype=mtype,
+                    name=name,
+                    value=value,
+                    sample_rate=sample_rate,
+                    tags=tags)
+
+    def _send_mtype(self):
+        """ send metric to sc using corrected
+            calls per metric type
+        """
+        func = getattr(sc, self.stat.mtype)
+        func_v2 = getattr(sc_v2, self.stat.mtype)
+
+        if self.stat.mtype == 'increment':
+            func(self.stat.name, self.stat.sample_rate, self.stat.tags)
+            func_v2(self.stat.name, self.stat.sample_rate, self.stat.tags)
+        elif self.stat.mtype == 'gauge' or self.stat.mtype == 'timing':
+            func(self.stat.name, self.stat.value, sample_rate=self.stat.sample_rate, tags=self.stat.tags)
+            func_v2(self.stat.name, self.stat.value, sample_rate=self.stat.sample_rate, tags=self.stat.tags)
+
+    def _send(self):
+        """ send metric to sc """
+        try:
+            self._send_mtype()
+        except Exception as error:
+            log.error('unable to send metric: {}'.format(error))
+
+    def _flush_cache(self):
+        """ read from cache, send every metric, truncate cache """
+        log.warning('flushing metrics from cache')
+        for stat in self.cache.read():
+            if not stat.deserialize():
+                # unable to parse stat, skip
+                log.error('unable to parse stat: {}'.format(stat))
+                continue
+            self.stat = stat
```

Comment:
pass stat as parameter to seld._send(stat) instead of defining member variable.

Reply:
The existing logic also reads from this member variable in other places as well, so changing this to a param means a lot of passing it around where it's needed